File: BasicDemo/bulletshooter.cpp
#include "bulletshooter.h"
#include "bulletworld.h"
#include <btBulletDynamicsCommon.h>


BulletShooter::BulletShooter()
    :
        m_shootBoxShape(0),
        m_ShootBoxInitialSpeed(40.f)
{
    btBoxShape* box = new btBoxShape(btVector3(.5f,.5f,.5f));
    box->initializePolyhedralFeatures();
    m_shootBoxShape = box;
}


BulletShooter::~BulletShooter()
{
	if (m_shootBoxShape)
		delete m_shootBoxShape;
}


void BulletShooter::shootBox(BulletWorld *world, 
        const btVector3 &camPos, const btVector3 &destination)
{
    btTransform startTransform;
    startTransform.setIdentity();
    //startTransform.setOrigin(camPos);

    btRigidBody* body = world->localCreateRigidBody(1.0f, startTransform, m_shootBoxShape);

    body->setLinearFactor(btVector3(1,1,1));

    body->getWorldTransform().setOrigin(camPos);
    body->getWorldTransform().setRotation(btQuaternion(0,0,0,1));
    btVector3 linVel(
            destination[0]-camPos[0],
            destination[1]-camPos[1],
            destination[2]-camPos[2]);
    linVel.normalize();
    linVel*=m_ShootBoxInitialSpeed;
    body->setLinearVelocity(linVel);
    body->setAngularVelocity(btVector3(0,0,0));
    body->setCcdMotionThreshold(0.5);
    body->setCcdSweptSphereRadius(0.9f);
}

File: swigbullet/python/examples/bulletdemo/bulletshooter.py
import swigbullet as bullet
from .vector3 import Vector3


class BulletShooter(object):
    def __init__(self):
        self.m_shootBoxShape=None
        self.m_ShootBoxInitialSpeed=40.0
        box = bullet.btBoxShape((0.5, 0.5, 0.5))
        box.initializePolyhedralFeatures()
        self.m_shootBoxShape=box

    def shootBox(self, world, camPos, destination):
        startTransform=bullet.btTransform()
        startTransform.setIdentity()

        body = world.localCreateRigidBody(1.0, startTransform, self.m_shootBoxShape)

        body.setLinearFactor((1.0,1.0,1.0))

        body.getWorldTransform().setOrigin(camPos);
        body.getWorldTransform().setRotation(bullet.btQuaternion(0,0,0,1));
        linVel=Vector3(
                destination[0]-camPos[0],
                destination[1]-camPos[1],
                destination[2]-camPos[2]).normalize()*self.m_ShootBoxInitialSpeed
        body.setLinearVelocity(linVel)
        body.setAngularVelocity((0.0, 0.0, 0.0))
        body.setCcdMotionThreshold(0.5)
        body.setCcdSweptSphereRadius(0.9)

